drivers/serial/uart_rtt.c

```diff
@@ -72,49 +86,152 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
 	SEGGER_RTT_Write(ch, &c, 1);
 }
 
+#ifdef CONFIG_UART_ASYNC_API
+
+static int uart_rtt_callback_set(struct device *dev, uart_callback_t callback,
+				 void *user_data)
+{
+	get_dev_data(dev)->callback = callback;
+	get_dev_data(dev)->user_data = user_data;
+	return 0;
+}
+
+static int uart_rtt_tx(struct device *dev, const uint8_t *buf, size_t len,
+		       int32_t timeout)
+{
+	const struct uart_rtt_config *cfg = get_dev_config(dev);
+	struct uart_rtt_data *data = get_dev_data(dev);
+	unsigned int ch = cfg ? cfg->channel : 0;
+
+	ARG_UNUSED(timeout);
+
+	/* RTT mutex cannot be claimed in ISRs */
+	if (k_is_in_isr()) {
+		return -ENOTSUP;
+	}
+
+	/* Claim the RTT lock */
+	if (k_mutex_lock(&rtt_term_mutex, K_NO_WAIT) != 0) {
```

Comment:
should you do `k_mutex_lock(&rtt_term_mutex, k_is_in_isr() ? K_NO_WAIT : K_FOREVER)`

Reply:
I would have thought so as well, but it is not valid to try locking a mutex inside an ISR at all, even with K_NO_WAIT.
https://github.com/zephyrproject-rtos/zephyr/blob/master/kernel/mutex.c#L125